DataFormats/L1Trigger/interface/P2GTCandidate.h

```diff
@@ -0,0 +1,328 @@
+#ifndef DataFormats_L1Trigger_P2GTCandidate_h
+#define DataFormats_L1Trigger_P2GTCandidate_h
+
+#include <vector>
+#include <ap_int.h>
+#include <stdexcept>
+
+#include "DataFormats/Common/interface/Ref.h"
+#include "DataFormats/Common/interface/RefVector.h"
+
+namespace l1t {
+
+  class L1GTProducer;
+
+  class P2GTCandidate;
+  typedef std::vector<P2GTCandidate> P2GTCandidateCollection;
+  typedef edm::Ref<P2GTCandidateCollection> P2GTCandidateRef;
+  typedef edm::RefVector<P2GTCandidateCollection> P2GTCandidateRefVector;
+  typedef std::vector<P2GTCandidateRef> P2GTCandidateVectorRef;
+
+  class P2GTCandidate {
+  public:
+    P2GTCandidate();
+
+    friend class L1GTProducer;
+
+    typedef ap_uint<16> hwPT_t;
+    typedef ap_int<13> hwPhi_t;
+    typedef ap_int<14> hwEta_t;
+    typedef ap_int<18> hwZ0_t;
+    typedef ap_uint<11> hwIso_t;
+    typedef ap_uint<8> hwQual_t;
+    typedef ap_uint<1> hwCharge_t;
+    typedef ap_int<12> hwD0_t;
+    typedef ap_uint<4> hwBeta_t;
+    typedef ap_uint<10> hwMass_t;
+    typedef ap_uint<16> hwIndex_t;
+    typedef ap_uint<10> hwSeed_pT_t;
+    typedef ap_int<10> hwSeed_z0_t;
+    typedef ap_uint<16> hwSca_sum_t;
+    typedef ap_uint<5> hwNumber_of_tracks_t;
+    typedef ap_uint<12> hwSum_pT_pv_t;
+    typedef ap_uint<2> hwType_t;
+    typedef ap_uint<8> hwNumber_of_tracks_in_pv_t;
+    typedef ap_uint<10> hwNumber_of_tracks_not_in_pv_t;
+
+    // Similar to std::optional<T> but avoids inheritance for ROOT file embedding
```

Comment:
What exactly is the problem with `std::optional`?

Reply:
https://root.cern/manual/io_custom_classes/#restrictions-on-types-root-io-can-handle
My educated guess is that this is because the implementation of `std::optional` uses multiple inheritance.

Reply:
@pcanal could you comment on ROOT IO and std::optional?

Reply:
ROOT I/O currently only support `std::optional` in `RNTuple`.  For other cases, if it works it is likely to not be platform independent.  There is plan to add support for `std::optional` in core I/O but we have not started the work yet.